cpp/modmesh/buffer/SimpleArray.hpp

```diff
@@ -725,6 +725,42 @@ class SimpleArrayMixinCalculators
         }
     }
 
+    A matmul(A const & other) const
+    {
+        auto athis = static_cast<A const *>(this);
+
+        if (athis->ndim() != 2 || other.ndim() != 2)
+        {
+            throw std::runtime_error("SimpleArray::matmul(): both arrays must be 2-dimensional");
```

Comment:
Use `std::out_of_range` for the index mismatch.  The error message should contain the shape of both `this` and `other`.  It may be convenient to use `modmesh::Formatter` to format the shape information in the error message.